FIP 0100 clarifications during last call

BigLep opened the pull request:
Starting to add clarifications that have emerged during public discussion in Last Call or while implementing.  I wanted a place to start putting these rather than just adding to https://github.com/filecoin-project/FIPs/discussions/1105 .  I don't expect to be the main author or contributor here.  I assume folks like @irenegia or @rvagg will do more here, or maybe discard this for a different PR.

BigLep commented:
(switched to draft since this isn't ready for review)

rvagg commented:
A lot of the content in here is mine so I'd prefer to skip on being a reviewer to get this merged.

momack2 commented on FIPS/fip-0100.md:
```diff
@@ -66,17 +64,23 @@ We propose a replacement mechanism that preserves these effects while adhering t
+More in details, for any sector onboarded after this FIP is deployed we have:
+
+- At ProveCommit time, we compute the per sector value as `dailyFee = k * CS(t) * sectorQAP`, where `k = 2.153 E-25` is a system constant,  `CS(t)` is the circulating supply value at the sector activation epoch, and `sectorQAP` is the quality adjusted power of the sector in bytes. The `dailyFee` value is stored but no payment is due at ProveCommit time.
+- Then, at the end of each deadline, we compute: `dailyPayment = min (deadline_dailyFee, m * expected_day_reward)`, where `deadline_dailyFee` is the sum of `dailyFee` of all sectors in the deadline, `m= 0.5` is another system constant and `expected_day_reward` is the updated value for the daily expected block reward for all sectors in the deadline. The `dailyPayment` value is burnt. Note that onl ...
```
```suggestion
- Then, at the end of each deadline, we compute: `dailyPayment = min (deadline_dailyFee, m * expected_day_reward)`, where `deadline_dailyFee` is the sum of `dailyFee` of all sectors in the deadline, `m= 0.5` is another system constant and `expected_day_reward` is the updated value for the daily expected block reward for all sectors in the deadline. The `dailyPayment` value is burnt. Note that only new[^1] sectors and old sectors extended or updated after this FIP is deployed count for the `deadline_dailyPayment` (indeed, any other sector has no `dailyFee` defined), while all sectors count for the `expected_day_reward` value.
```

momack2 reviewed (approved):
A few tiny nits - solid additional implementation detail! 👍

rvagg commented:
Latest [commit](https://github.com/filecoin-project/FIPs/pull/1124/commits/d94b4d1f96ee28d1da4ed87489fbb77db6cc878c) on here increases the accuracy of the per-byte multiplier from `2.153e-25` to `2.1537e-25`. During implementation I noted that the extra dp made enough of a difference in getting closer to `7.4e-15` to make my test numbers look significantly closer to the previous versions that had `7.4e-15` and no QAP. e.g. in a test case that's close to current mainnet conditions with 680M CS and a 32GiB CC sector it's `0.000005032` (FIL) for the straight `7.4e-15` multiplier and `0.000005032038659574` with the per-byte form vs `0.000005030403136028` without that additional dp.

BigLep commented on FIPS/fip-0100.md:
````diff
@@ -144,27 +184,61 @@ pub struct SectorOnChainInfo {
+}
+```
+
+**Sector commitments**: At each of `ProveCommitSectors3`, `ProveCommitAggregate` and `ProveCommitSectorsNI`:
````
Here is an example: your call on whether you think this is better.  Feel free to skip:

Action | `SectorOnChainInfo.daily_fee` handling for new sectors | `SectorOnChainInfo.daily_fee` handling for old sectors | `Deadline.daily_fee` handling | `ExpirationSet.fee_deduction` handling
-- | -- | -- | -- | --
Sector commitments<br/>(`ProveCommitSectors3`, `ProveCommitAggregate`, `ProveCommitSectorsNI`) | Calculated based on the circulating supply at that epoch and the QAP | n/a | Increased by `daily_fee` as sectors are assigned to deadlines. | Increased by `daily_fee` as sectors are further assigned to partitions within the deadline and their expiration recorded in the partition's expiration queue.
Replica updates (snaps)<br/>(`ProveReplicaUpdates`, `ProveReplicaUpdates3`) | Uses the circulating supply value at the time the sector was onboarded. The new `daily_fee` value is changed in proportion to the change in the sector's quality-adjusted power (i.e. `daily_fee *= new_QAP / old_QAP`). | Its value will be calculated according to the circulating supply at the time of the update and the new quality-adjusted power of the sector. | Same as above. | Same as above.
Sector extensions | No change since extensions don't change a sector's QAP but only the duration of the sector. | Its value will be calculated according to the circulating supply at the time of the extension and the sector's QAP. | Same as above. | Same as above.
Faults | None | n/a | None | Faults can cause changes to a partition's expiration queue to ensure proper handling of a faulty sector's lifecycle. Changes in `ExpirationSet` to move sectors between queues will also include corresponding changes to `fee_deduction` values.
Early terminations (automatic and manual) | None | n/a | Reduced by `SectorOnChainInfo.daily_fee` |  

rvagg replied:
Good idea, I've adapted this to a summary table at the bottom of this section that I think gives a terse summary of actions and change states.

BigLep reviewed (commented):
Nice work. Lots of great updates!

The only thing from https://github.com/filecoin-project/FIPs/discussions/1105#discussioncomment-12194371 that I din't see covered was "Given fee payments will happen in each Deadline and the cap calculation will use Deadline.live_power ([reference](https://github.com/filecoin-project/FIPs/blob/ea103d897cfe16ad0fd50a05fab91466c05997e3/FIPS/fip-0100.md?plain=1#L119)), this means pre-FIP-0100 sectors will affect the cap."

rvagg commented on FIPS/fip-0100.md:
```diff
@@ -66,17 +64,23 @@ We propose a replacement mechanism that preserves these effects while adhering t
  - Enable future optimizations in the network’s execution layer.
 
-**Proposal Details**: We introduce a per-sector fee, whose value is proportional to a fixed fraction of the circulating supply and to the sector duration. To prevent this fee from becoming a significant upfront cost that could hinder onboarding, we propose daily payments instead of an upfront fee. Moreover, as a safety measure for the high-growth scenarios, we propose capping the daily payment at ...
+**Proposal Details**: We introduce a per-sector fee, whose value is proportional to a fixed fraction of the circulating supply, to the sector duration and to the sector QAP (Quality Adjusted Power). To prevent this fee from becoming a significant upfront cost that could hinder onboarding, we propose daily payments instead of an upfront fee. Moreover, as a safety measure for the high-growth scenar ...
```
implied, expected block reward is a function of a particular sector so it accounts for that sector's power; I've added some clarifying language here though

rvagg commented on FIPS/fip-0100.md:
````diff
@@ -112,28 +116,64 @@ Aggregate fee functionality and constants will be removed from the storage miner
+}
+```
+
+A daily fee value is calculated per-sector at time of onboarding and whenever a sector's power is adjusted based on circulating supply and power at each point. The fee value is stored as `daily_fee` in the `SectorOnChainInfo` and an aggregated `daily_fee` value is also stored in each `Deadline`. The fee is deducted from the miner actor's balance, processed by cron at the end of each deadline.
````
good catch

rvagg commented on FIPS/fip-0100.md:
```diff
@@ -229,30 +303,30 @@ Filecoin node implementations compatible with Calibnet will be required to set t
 <div align="center">
     <img src="https://github.com/user-attachments/assets/018fb429-de90-4a86-9c15-6c244a699487">
-    <p><em>Figure 2: CS-based fee evolution as a function of various network metrics. Here we considered a daily payment that is a fixed % of CS-based and capped at 50% of `expected_day_reward` as a safety measure in high-growth scenarios.
+    <p><em>Figure 2: CS-based fee evolution as a function of various network metrics. Here we considered a daily payment that is a fixed % of CS-based and capped at 50% of `expected_day_reward` as a safety measure in high-growth scenarios. The Fee/Sector values assume 32GiB of QAP. 
```
I've changed it to this:

> The Fee/Sector values assume 32GiB of QAP, i.e. a simplified modelling where sectors are all 32GiB CC.

irenegia commented:
@BigLep 
> Nice work. Lots of great updates!
> 
> The only thing from [#1105 (comment)](https://github.com/filecoin-project/FIPs/discussions/1105#discussioncomment-12194371) that I din't see covered was "Given fee payments will happen in each Deadline and the cap calculation will use Deadline.live_power ([reference](https://github.com/filecoin-project/FIPs/blob/ea103d897cfe16ad0fd50a05fab91466c05997e3/FIPS/fip-0100.md?plain=1#L119)), this means pre-FIP-0100 sectors will affect the cap."

See "Note that only new[2](https://github.com/filecoin-project/FIPs/blob/fip-0100/clarifications-during-last-call/FIPS/fip-0100.md#user-content-fn-1-db6c88c5768104c115309060f0ef1366) sectors and old sectors extended or updated after this FIP is deployed count for the deadline_dailyFee (indeed, any other sector has no dailyFee defined), while all sectors count for the expected_day_reward value." in https://github.com/filecoin-project/FIPs/blob/fip-0100/clarifications-during-last-call/FIPS/fip-0100.md#per-sector-fee-added

BigLep commented:
> any other sector has no

I see there is also "Note that because live_power is a sum of all deadline power and will include pre-FIP-0100 sectors without a daily_fee as well as new and updated sectors with a daily_fee, the cap will account for all sectors in the deadline, while the fee that is being capped will only be for the new and updated sectors, deriving a higher cap in the early period of activation of this FIP than would be expected once all sectors in the system have a daily_fee value."

It seems well documented now - good stuff

BigLep commented:
Great updates @irenegia and @rvagg!  It looks like @kkarrancsu is the main reviewer want to get confirmation from.  After that will we merge?  If another round of reminding the community that this is in "last call" is done that it would be ideal to point them to https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0100.md directly if possible rather than this PR (which also has another PR stacked on top of it).

BigLep commented:
@luckyparadise : I think at this point it's best to move forward with merging this PR so that https://github.com/filecoin-project/FIPs/blob/master/FIPS/fip-0100.md can be referenced for the community to come up to speed on the FIP since a lot of good clarifications have been added with this PR.  Given it has received approval from a FIP editor, can you please merge?
